### PR TITLE
test(e2e): add test case for two bi's with duplicate resources

### PR DESCRIPTION
**Pull request by tylerslaton**
## Summary
Adding a e2e test case for creating two BundleInstances that reference the same Bundle. In this particular case we are simulating and expecting a InstallFailed status with a Message of that references the duplicated resources. 

## Open Questions
1. Is there a situation where we want to have multiple BundleInstances reference the same Bundle? If so, how do we get around this test in the future?
2. Is this a valid way of checking this behavior?

### CODE REVIEW

File: test/e2e/plain-v0_provisioner_test.go [hunk 1]
```diff
@@ -423,8 +423,118 @@ var _ = Describe("plain-v0 provisioner bundleinstance", func() {
...
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "olm-apis-duplicate",
+				},
+				Spec: rukpakv1alpha1.BundleInstanceSpec{
+					ProvisionerClassName: plainProvisionerID,
+					BundleName:           bundle.GetName(),
+				},
+			}
+
+			err = c.Create(ctx, biDuplicate)
```
Comment (exdx) [inline test/e2e/plain-v0_provisioner_test.go:473]: I'm not sure but I think this may be a bit racy. Creating one object after the other sequentially like this doesn't necessarily mean that they will be created on the API server in the same order. I guess for a small kind cluster with no load we can expect it to be the case, but for example on Openshift it may not be always true that the second bundle is always seen after the first bundle if you create them at the same time. Maybe we could add an Eventually() or even sleep between creating the first and second BI, just to ensure the test is not prone to these types of timing issues?

**Review by exdx — COMMENTED** [review]
Nice test. I don't think there is a scenario where we would want two BI's referencing the same Bundle in the new cluster-scoped operator world that Rukpak is based on. In OLM today, operators can install themselves in a specific namespace and watch a set of other namespaces. This behavior supports multitenancy, where multiple copies of the same operator can exist on the same cluster. The idea is to move away from this model to one where every operator (BI) is by default cluster-scoped and no more than one version should ever be running on a cluster at a given time. 

We could prevent this by keeping in memory a map of BIs -> BundleNames, and reference that before installing a new BI. This map would have to stay up to date in the case of deletion events though, and may slowly drift from the state of the cluster. Another option is an admission webhook that intercepts the BI create request, and queries the state of the cluster to see if there is already a BI referencing that bundle. Third option would be to simply let the new BI fail (as we do now) and provide a really clear error message as to why it happened

File: test/e2e/plain-v0_provisioner_test.go (hunk 1 above)
Comment (tylerslaton) [reply test/e2e/plain-v0_provisioner_test.go:473]: What if we had an Eventually stop gap between them, like you said, that attempts to grab each BI. It would guarantee that it is created and thus prevent proceeding before both are existing and ready.

File: test/e2e/plain-v0_provisioner_test.go [hunk 2]
```diff
@@ -423,8 +423,118 @@ var _ = Describe("plain-v0 provisioner bundleinstance", func() {
...
+	})
 })
 
 func conditionsSemanticallyEqual(a, b metav1.Condition) bool {
 	return a.Type == b.Type && a.Status == b.Status && a.Reason == b.Reason && a.Message == b.Message
 }
+
+func conditionsLooselyEqual(a, b metav1.Condition) bool {
+	return a.Type == b.Type && a.Status == b.Status && a.Reason == b.Reason && strings.Contains(b.Message, a.Message)
+}
```
Comment (timflannagan) [inline test/e2e/plain-v0_provisioner_test.go:540]: Yup - this is exactly why I think 1:1 error message (or in this case status condition message) checking makes tests brittle over time. I think this is fine for now, but we really should move towards a strings.Contains approach imo.

File: test/e2e/plain-v0_provisioner_test.go [hunk 3]
```diff
@@ -423,8 +423,118 @@ var _ = Describe("plain-v0 provisioner bundleinstance", func() {
...
+				looselyExpectedMessage := "rendered manifests contain a resource that already exists"
+				expected := metav1.Condition{
+					Type:    "Installed",
+					Status:  metav1.ConditionStatus(corev1.ConditionFalse),
+					Reason:  "InstallFailed",
+					Message: looselyExpectedMessage,
+				}
+
+				return conditionsLooselyEqual(expected, *existing)
+			}).Should(BeTrue())
```
Comment (timflannagan) [inline test/e2e/plain-v0_provisioner_test.go:529]: not blocking: it would be nice to generalize this now that we have this pattern in a couple of different places.

File: test/e2e/plain-v0_provisioner_test.go (hunk 1 above)
Comment (timflannagan) [reply test/e2e/plain-v0_provisioner_test.go:473]: I think this is fine for now imo but Dan's point seems valid - we may need a couple of different helper functions to drive Bundle/BI resources creation at this point, and revisit refactoring the e2e suite structure I had built up.

**Review by timflannagan — APPROVED** [review]

File: test/e2e/plain-v0_provisioner_test.go (hunk 1 above)
Comment (tylerslaton) [reply test/e2e/plain-v0_provisioner_test.go:473]: I'll make a todo comment for it 👍🏻

File: test/e2e/plain-v0_provisioner_test.go (hunk 2 above)
Comment (tylerslaton) [reply test/e2e/plain-v0_provisioner_test.go:540]: So you're thinking instead of having exact equality, all of these should be `.Contains`?

File: test/e2e/plain-v0_provisioner_test.go (hunk 3 above)
Comment (tylerslaton) [reply test/e2e/plain-v0_provisioner_test.go:529]: I agree, I have a solution coming out soon for this that should clean that up.

File: test/e2e/plain-v0_provisioner_test.go (hunk 1 above)
Comment (tylerslaton) [reply test/e2e/plain-v0_provisioner_test.go:473]: https://github.com/operator-framework/rukpak/issues/138
